daisy/step_create_images.go

```diff
@@ -16,63 +16,124 @@ package daisy
 
 import (
 	"context"
+	"encoding/json"
 	"sync"
 
 	"google.golang.org/api/googleapi"
 )
 
 // CreateImages is a Daisy CreateImages workflow step.
-type CreateImages []*Image
+type CreateImages struct {
+	Images     []*Image
+	ImagesBeta []*ImageBeta
+}
+
+// UnmarshalJSON unmarshals Image.
+func (ci *CreateImages) UnmarshalJSON(b []byte) error {
+	imagesBeta := &[]*ImageBeta{}
+	if err := json.Unmarshal(b, imagesBeta); err != nil {
+		return err
+	}
+	ci.ImagesBeta = *imagesBeta
+
+	images := &[]*Image{}
+	if err := json.Unmarshal(b, images); err != nil {
+		return err
+	}
+	ci.Images = *images
+
+	return nil
+}
+
+func usesBetaFeatures(imagesBeta []*ImageBeta) bool {
+	for _, imageBeta := range imagesBeta {
+		if imageBeta != nil && len(imageBeta.StorageLocations) > 0 {
+			return true
+		}
+	}
+	return false
+}
 
 // populate preprocesses fields: Name, Project, Description, SourceDisk, RawDisk, and daisyName.
 // - sets defaults
 // - extends short partial URLs to include "projects/<project>"
-func (c *CreateImages) populate(ctx context.Context, s *Step) dErr {
+func (ci *CreateImages) populate(ctx context.Context, s *Step) dErr {
 	var errs dErr
-	for _, i := range *c {
-		errs = addErrs(errs, i.populate(ctx, s))
+	if ci.Images != nil {
+		for _, i := range ci.Images {
+			errs = addErrs(errs, populate(ctx, i, &i.ImageBase, s))
+		}
 	}
+
+	if ci.ImagesBeta != nil {
+		for _, i := range ci.ImagesBeta {
+			errs = addErrs(errs, populate(ctx, i, &i.ImageBase, s))
+		}
+	}
+
 	return errs
 }
 
-func (c *CreateImages) validate(ctx context.Context, s *Step) dErr {
+func (ci *CreateImages) validate(ctx context.Context, s *Step) dErr {
 	var errs dErr
-	for _, i := range *c {
-		errs = addErrs(errs, i.validate(ctx, s))
+
+	if usesBetaFeatures(ci.ImagesBeta) {
+		if ci.ImagesBeta != nil {
```

Comment:
These nil checks right before the range are not needed.

Reply:
Removed